Fix various non-critical issues

kerneltoast opened the pull request.

kerneltoast committed:
lab4: cache: Fix parameter ordering in the Cache constructor

The ordering in the parameter list should match the ordering in the class
definition. This didn't cause any problems here, but if there were
dependencies between the different parameters for their initialization,
then the incorrect initialization ordering could have caused undefined
behavior.

kerneltoast committed:
lab4: processor: Add missing field to push() calls in createRecords()

struct Record contains six parameters, not five. The sixth parameter should
be explicitly initialized to a null pointer here.

kerneltoast committed:
lab4: Add missing header guards

All of the other headers have them; we should be consistent.

kerneltoast committed:
lab4: main: Add missing return to main()

kerneltoast committed:
lab4: processor: Use memset() in sendRequest() and fix inconsistency

The bitwise AND with 0xFF is unnecessary when casting to uint8_t, so
remove it. Also use a memset here since it's equivalent to what this loop
is doing (and it does it faster).

Additionally, clean up the inconsistency with the arguments being passed to
cache->receiveRequest() (all of the arguments are members of Record r).

There's no need to set r.data to null when not writing either since it
defaults to null.

This commit does not alter the functionality of processor.cc; everything
still works as expected.

powerjg commented on lab4/provided/processor.cc:
```diff
@@ -20,19 +21,15 @@ Processor::~Processor()
-            data[i] = (uint8_t)(r.requestId & 0xFF);
-        }
+        r.data = new uint8_t[r.size];
+        memset(r.data, (uint8_t)r.requestId, r.size);
```
The memset change is good here, but I'd rather not push the other changes. I'm planning on adding some testing code that will use data soon.

powerjg replied:
I've made some updates here now. Could you look at merging with the current master and see if this still makes sense. If so, I'm happy to merge it.

kerneltoast replied:
The updates you made didn't affect this; it looks good to me.

powerjg commented on lab4/provided/Makefile:
```diff
@@ -1,3 +1,33 @@
+
+DEPFLAGS = -MMD -MF $(@:.o=.d)
+deps := $(patsubst %.o,%.d,$(objs))
+-include $(deps)
```
This is more makefile magic than I know...

animalvgamer replied:
This looks okay actually, the lines above just quickly set some dependency file references. Honestly my only bad comment is that the -g and debug flags should be kept separate from the main flags, and adjustable by Makefile argument.

powerjg replied:
+1, good idea.

kerneltoast replied:
Done @animalvgamer.

powerjg reviewed (commented):
Hi Sultan,

Most of the changes look good to me. One minor thing that I would rather not merge, though.

Thanks!!

kerneltoast committed:
lab4: Make the Makefile fancy

Now, make will catch changes made to individual files and will recompile
only the necessary components of cache_simulator, instead of requiring
the user to delete cache_simulator and compile every file again every time
a file is changed.

A clean rule has been added as well, and debugging is disabled by default.

powerjg commented:
Thanks for the updates!
